core/drivers/lpc_uart.c

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, HiSilicon Limited
+ */
+
+#include <assert.h>
+#include <drivers/lpc_uart.h>
+#include <io.h>
+#include <keep.h>
+#include <mm/core_mmu.h>
+#include <util.h>
+
+static paddr_t chip_to_base(struct serial_chip *chip)
+{
+	struct lpc_uart_data *pd =
+		container_of(chip, struct lpc_uart_data, chip);
+
+	return io_pa_or_va(&pd->base, LPC_SIZE);
+}
+
+static void lpc_byte_read(paddr_t addr, unsigned char *data)
+{
+	unsigned int cnt = 0;
+	unsigned int status;
+
+	io_write32((LPC_CMD_REG_OFFSET) + addr, 0x8);
+
+	io_write32((LPC_OP_LEN_REG_OFFSET) + addr, 1);
+	io_write32((LPC_ADDR_REG_OFFSET) + addr, UART_BASE + UART_LSR);
+
+	io_write32((LPC_START_REG_OFFSET) + addr, 1);
+
+	status = io_read32((LPC_IRQ_ST_REG_OFFSET) + addr);
+	while (!(status & 0x2)) {
+		if (cnt > UART_SEND_LOOP_MAX)
+			return;
+		cnt++;
+		status = io_read32((LPC_IRQ_ST_REG_OFFSET) + addr);
+	}
+
+	io_write32((LPC_IRQ_ST_REG_OFFSET) + addr, 0x2);
+
+	if (io_read32((LPC_OP_STATUS_REG_OFFSET) + addr) & 0x2)
+		*data = io_read32((LPC_RDATA_REG_OFFSET) + addr);
+}
+
+static void lpc_byte_write(paddr_t addr, unsigned char data)
+{
+	unsigned int cnt = 0;
+	unsigned int status;
+
+	io_write32((LPC_CMD_REG_OFFSET + addr), 0x9);
+	io_write32((LPC_OP_LEN_REG_OFFSET + addr), 1);
+	io_write32((LPC_WDATA_REG_OFFSET + addr), data);
+
+	io_write32((LPC_ADDR_REG_OFFSET + addr), UART_BASE + UART_THR);
+	io_write32((LPC_START_REG_OFFSET + addr), 1);
+
+	status = io_read32((LPC_IRQ_ST_REG_OFFSET + addr));
+	while (!(status & 0x2)) {
+		if (cnt > UART_SEND_LOOP_MAX)
+			return;
+		cnt++;
+		status = io_read32((LPC_IRQ_ST_REG_OFFSET + addr));
+	}
+
+	io_write32((LPC_IRQ_ST_REG_OFFSET + addr), 0x2);
+}
+
+static void lpc_uart_core_putc(paddr_t base, int ch)
+{
+	unsigned int loop = 0;
+	unsigned char var = '\0';
+
+	while (loop < UART_SEND_LOOP_MAX) {
+		lpc_byte_read(base, &var);
+
+		if ((var & 0x40) == 0x40)
+			break;
+		loop++;
+	}
```

Comment:
Please use a `for` loop, and same below.

Reply:
done